extensions/replication/utils/ReplicationTaskScheduler.js

```diff
@@ -1,24 +1,63 @@
 const async = require('async');
 
+const ObjectQueueEntry = require('../../../lib/models/ObjectQueueEntry');
+const ActionQueueEntry = require('../../../lib/models/ActionQueueEntry');
+
 /**
- * Schedule replication tasks for replicated buckets, by ensuring that
- * all updates to the same object are serialized
-*/
+ * Schedule replication tasks for replication tasks, by ensuring that all
+ * updates to the same object are serialized
+ */
 class ReplicationTaskScheduler {
     constructor(processingFunc) {
         this._processingFunc = processingFunc;
         this._runningTasksByMasterKey = {};
+        this._runningTasks = {};
+    }
+
+    _getUniqueKey(ctx) {
+        const { entry } = ctx;
+        let key;
+        let version;
+        let contentMd5;
+        if (entry instanceof ObjectQueueEntry) {
+            key = entry.getCanonicalKey();
+            version = entry.getVersionId();
+            contentMd5 = entry.getContentMd5();
+        }
+        if (entry instanceof ActionQueueEntry) {
+            ({ key, version, contentMd5 } = entry.getAttribute('target'));
+        }
+        return `${key}:${version || ''}:${contentMd5}`;
+    }
+
+    _queueCallback(ctx, done) {
+        delete this._runningTasks[this._getUniqueKey(ctx)];
+        done();
+    }
+
+    _isRunningTask(ctx) {
+        return this._runningTasks[this._getUniqueKey(ctx)];
+    }
+
+    _getNewQueue(ctx, masterKey) {
+        const queue = async.queue(this._processingFunc);
+        queue.drain = () => delete this._runningTasksByMasterKey[masterKey];
+        this._runningTasksByMasterKey[masterKey] = queue;
+        this._runningTasks[this._getUniqueKey(ctx)] = true;
```

Comment:
I think this update is misplaced. It should be done on *every* new incoming task that is not skipped, including those that get queued to an existing queue.

To test this you would have to add 3 entries:

- two with the same master key but different version and/or md5 (second entry should be queued after the first)
- one with the same master/version/md5 than the second entry

Then expect the latest entry to be skipped (I think in the current situation it would not be)

Reply:
👍 Updated and added this test case.